nautobot/extras/tests/test_dynamicgroups.py

```diff
@@ -0,0 +1,360 @@
+        # filter we set from the form should be identical to what was there already.
+        old_filter = group.filter
+        form_class = group.generate_filter_form()
+        form_class.prefix = None  # We don't want the prefix here right now.
+        form = form_class(group.filter)
+        form.is_valid()
+        group.set_filter(form.cleaned_data)
+        self.assertEqual(group.filter, old_filter)
+
+        # Now we'll do it using a manually crafted dict.
+        new_filter = {"interfaces": True}
+        group.set_filter(new_filter)
+        self.assertEqual(group.filter, new_filter)
+
+        # And a bad input
+        bad_filter = {"site": -42}
+        with self.assertRaises(ValidationError):
+            group.set_filter(bad_filter)
+
+        # Cleanup
+        group.refresh_from_db()
+
+
+class DynamicGroupMapTest(DynamicGroupTestBase):
+    def test_dynamicgroup_map_factory(self):
+        """Test `dynamicgroup_map_factory()`."""
+        group = self.groups[0]
+        model = group.content_type.model_class()
+        dynamicgroup_map = dynamicgroup_map_factory(model)
+        self.assertEqual(dynamicgroup_map.model, model)
+        self.assertEqual(dynamicgroup_map.filterset_class, DeviceFilterSet)
+        self.assertEqual(dynamicgroup_map.filterform_class, DeviceFilterForm)
+        self.assertEqual(dynamicgroup_map.form_class, DeviceForm)
+
+        # Bogus input returns None.
+        self.assertIs(dynamicgroup_map_factory(False), None)
+
+    def test_base_url(self):
+        """Test `DynamicGroup.map.base_url`."""
+        group = self.groups[0]
+        self.assertEqual(group.map.base_url, reverse("dcim:device_list"))
+
+    def test_fields(self):
+        """Test `DynamicGroup.map.fields()`."""
+        group = self.groups[0]
+        fields = group.map.fields()
+
+        # Test that it's a dict with or without certain key fields.
+        self.assertNotIn("q", fields)
+        self.assertIn("name", fields)
+        self.assertIsInstance(fields, dict)
+        self.assertNotEqual(fields, {})
+
+    def test_get_queryset(self):
+        """Test `DynamicGroup.map.get_queryset()`."""
+        group = self.groups[0]
+        device1 = self.devices[0]
+
+        # Test that we can get a full queryset
+        qs = group.map.get_queryset(group.filter)
+        devices = group.content_type.model_class().objects.filter(site=device1.site)
+        self.assertEqual(list(qs), list(devices))
+
+        # ... or just a list of UUIDs
+        qs_flat = group.map.get_queryset(group.filter, flat=True)
+        devices_flat = devices.values_list("pk", flat=True)
+        self.assertEqual(sorted(qs_flat), sorted(devices_flat))
+
+    def test_urlencode(self):
+        """Test `DynamicGroup.map.urlencode()`."""
+        group = self.groups[0]
+        params = group.map.urlencode(group.filter)
+        expected = "site=site-1"
+        self.assertEqual(params, expected)
+
+
+class DynamicGroupFilterTest(DynamicGroupTestBase):
```

Comment:
Ideally I'd think this should be in `nautobot/extras/tests/test_filters.py`?

Reply:
Yeah I thought to just put all the`DynamicGroup` tests here since they are all using the same base. The alternative would be to import that base in the other test modules? Just trying to reduce duplication if at all possible.

Reply:
Yeah, this is fine as is. :-)